elyra/pipeline/validate.py

```diff
@@ -0,0 +1,633 @@
+from elyra.pipeline.component_registry import ComponentRegistry
+from elyra.pipeline.parser import PipelineParser
+from elyra.pipeline.pipeline import Operation
+from elyra.pipeline.processor import PipelineProcessorManager
+from elyra.util.path import get_expanded_path
+
+
+class ValidationSeverity(IntEnum):
+    Error = 1
+    Warning = 2
+    Information = 3
+    Hint = 4
+
+
+class ValidationResponse(object):
+    def __init__(self):
+        self._response = {"title": "Elyra Pipeline Diagnostics",
+                          "description": "Issues discovered when parsing the pipeline",
+                          "issues": []
+                          }
+        self._has_fatal = False
+
+    @property
+    def response(self) -> Dict:
+        """
+        :return: The dict of validation errors and warnings found in the pipeline
+        """
+        return self._response
+
+    @property
+    def has_fatal(self):
+        return self._has_fatal
+
+    def add_message(self,
+                    message: str,
+                    message_type: Optional[str] = "",
+                    data: Optional[Dict] = "",
+                    severity: ValidationSeverity = ValidationSeverity.Warning):
+        """
+        Helper function to add a diagnostic message to the response to be sent back
+        :param message: A simple message describing the issue
+        :param message_type: The type of message to send back e.g. invalidNodeType, invalidPipeline
+        :param data: a Dict with granular details regarding the error e.g. the nodeID, pipelineID, linkID etc.
+        :param severity: the severity level of the issue
+        :return:
+        """
+        valid_severity_levels = [ValidationSeverity.Error, ValidationSeverity.Warning,
+                                 ValidationSeverity.Information, ValidationSeverity.Hint]
+
+        if severity in valid_severity_levels:
+            diagnostic = {"severity": severity.value,
+                          "source": "Elyra Pipeline Validation Service",
+                          "type": message_type,
+                          "message": message,
+                          "data": data
+                          }
+            self._response['issues'].append(diagnostic)
+
+        if severity is ValidationSeverity.Error:
+            self._has_fatal = True
+
+    def to_json(self):
+        return self._response
+
+
+class PipelineValidationManager(SingletonConfigurable):
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.root_dir = get_expanded_path(kwargs.get('root_dir'))
+
+    async def validate(self, pipeline: Dict) -> ValidationResponse:
+        """
+        Validates the pipeline JSON payload
+        :param pipeline: the pipeline definition to be validated
+        :return: ValidationResponse containing any and all issues discovered during the validation
+        """
+
+        response = ValidationResponse()
+        root_dir = self.root_dir
```

Comment:
This is not necessary since it's not used (after updating argument list of `self._validate_node_properties()`)

Reply:
fixed